### PR TITLE
proposal: Add logging standard

### LINKED ISSUE #6: [Proposal]: Introduce a logging standard
### Describe the change

Currently all Zupit applications have either no logging, or if they have it, it is up to the team / developer on what and how to log.

With a standard, we could make sure that each application developed in Zupit respects a standard and actually logs important (but not sensitive!) informations.		

### Describe how this change might impact positively Zupit

- Logs are actually useful and log the desired informations
- All applications have logs (Both front-end and back-end)
- With good logging, it is easier for us debug errors

### Describe how this change might impact negatively Zupit

- It might impact negatively performances of teams when introducing the standard.
- There might be some corner-cases where logging must be done differently

### PR DESCRIPTION
**Pull request by alessandro-mariotti-zupit**
Closes #6

### CODE REVIEW

File: common/logging.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
+# Applications Logging
```
Comment (alessandro-mariotti-zupit) [inline common/logging.md:1]: Created Issue for improving this https://github.com/zupit-it/zupit-standards/issues/17

File: common/logging.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+## Logging platforms
+
+Where we can collect, query, and organize logs from apps.
+
+We use [Application Insights](https://learn.microsoft.com/en-us/azure/azure-monitor/app/app-insights-overview?tabs=net) (an extension of [Azure monitor](https://learn.microsoft.com/en-us/azure/azure-monitor/overview)) for logging.
+
+There could be some exceptions on this, based on customer's requirements or project constraints, where we fallback on ad-hoc solutions or classic logfiles (see the next sections for details)
+
+
+## Code for Logging
```
Comment (alessandro-mariotti-zupit) [inline common/logging.md:23]: Created issue for improving this: https://github.com/zupit-it/zupit-standards/issues/18

File: common/logging.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+### Hierarchy
+
+Loggers should reflect the hierarchy of the application, so that logging level can be changed for narrowed sections of the application in an easy way  
+
+For example:
+
+```
+- root
+ `- api
```
Comment (omar-muscatello-zupit) [inline common/logging.md:58]: Why are this backticks needed?

File: common/logging.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+Where possible, it should be possible to change the logging level at Runtime. 
+
+## What to log
+
+Each log entry should contain the following informations:
+
+- Timestamp
+- Thread / Process *(Where available)*
+- Logged user *(Where available)*
```
Comment (omar-muscatello-zupit) [inline common/logging.md:83]: Is it referring to the user id? is it safe to log it?

File: common/logging.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+- Massimo Telch (massimo.telch@zupit.it)
+- Pietro De Vigili (pietro.devigili@zupit.it)
+
+## Logging platforms
+
+Where we can collect, query, and organize logs from apps.
+
+We use [Application Insights](https://learn.microsoft.com/en-us/azure/azure-monitor/app/app-insights-overview?tabs=net) (an extension of [Azure monitor](https://learn.microsoft.com/en-us/azure/azure-monitor/overview)) for logging.
+
+There could be some exceptions on this, based on customer's requirements or project constraints, where we fallback on ad-hoc solutions or classic logfiles (see the next sections for details)
```
Comment (omar-muscatello-zupit) [inline common/logging.md:20]: `logfiles` typo?

File: common/logging.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+- App version *(Where available)*
+
+When logging an exception, always log the related stacktrace.
+
+The application should log all useful information to recreate locally the scenario that caused the bug and/or error. Here are some guidelines:
+
+### Endpoint/API Layer
+
+- Every Endpoint/API must log that it was called, from whom (logged user), what it does, it's input parameters (Path variables and query params) with **INFO LEVEL** 
+- Every Endpoint/API must log it's request body (if any) with **DEBUG LEVEL**
```
Comment (omar-muscatello-zupit) [inline common/logging.md:94]: I don't think this is safe. For example, when calling the authentication endpoint, the request body would include user credentials (e.g. OAuth Resource Owner Flow).

We should also consider GDPR compliance: request body could include user information (SaveProfile API).

This comment is not valid if we should have different log levels for each environment. For example:
- Testing -> DEBUG: even if we log user credentials, those are testing credentials

What do you think?

File: common/logging.md (hunk 3 above)
Comment (alessandro-mariotti-zupit) [reply common/logging.md:58]: Because it's a directory tree structure and  I like how it looks

File: common/logging.md [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+## Owners
+
+For this standard, you can refer to:
+
+- Alessandro Mariotti (alessandro.mariotti@zupit.it)
+- Alessandro Saiani (alessandro.saiani@zupit.it)
+- Massimo Telch (massimo.telch@zupit.it)
+- Pietro De Vigili (pietro.devigili@zupit.it)
+
```
Comment (pietro-devigili-zupit) [inline common/logging.md:13]: Add Omar to list?

**Review by pietro-devigili-zupit — CHANGES_REQUESTED** [review]

**Review by pietro-devigili-zupit — APPROVED** [review]

**Review by pietro-devigili-zupit — APPROVED** [review]

**Review by omar-muscatello-zupit — APPROVED** [review]

**Review by alessandro-saiani-zupit — APPROVED** [review]

**Pull request merged**